lib/translator/inquiry.cc

```diff
@@ -0,0 +1,265 @@
+  char product_serial_number[40] = {0}, hex_string[33];
+
+  // TOOD: Create shared constants to be used in tests as well when merging all
+  // Inquiry paths
+  const size_t kNguidLen = 40, kEui64Len = 20, kV1SerialLen = 30;
+
+  // check if nonzero
+  bool nguid_nz = identify_ns.nguid[0] || identify_ns.nguid[1];
+  bool eui64_nz = identify_ns.eui64;
+
+  if (nguid_nz || eui64_nz) {
+    if (nguid_nz) {
+      // 6.1.3.1.1
+      result.page_length = kNguidLen;
+
+      // convert 128-bit number into hex string, 64-bits at a time
+      sprintf(hex_string, "%08lx", identify_ns.nguid[0]);
+      sprintf(&hex_string[16], "%08lx", identify_ns.nguid[1]);
+
+    } else if (eui64_nz) {
+      // 6.1.3.1.2
+      result.page_length = kEui64Len;
+
+      // convert 64-bit number into hex string
+      sprintf(hex_string, "%08lx", identify_ns.eui64);
+    }
+
+    // insert _ and . in the correct positions
+    for (int i = 4; i < result.page_length - 1; i += 5) {
+      product_serial_number[i] = '_';
+    }
+    product_serial_number[result.page_length - 1] = '.';
+
+    // insert numbers
+    int pos = 0;
+    for (int i = 0; i < result.page_length - 1; ++i) {
+      if (product_serial_number[i] != '_') {
+        product_serial_number[i] = hex_string[pos++];
+      }
+    }
+  } else {
+    // 6.1.3.1.3
+    // valid for NVMe 1.0 devices only
+
+    result.page_length = kV1SerialLen;
+
+    // PRODUCT SERIAL NUMBER should be formed as follows:
+    // Bits 239:80 20 bytes of Serial Number (bytes 23:04 of Identify
+    // Controller data structure)
+    static_assert(sizeof(identify_ctrl.sn) == 20);
+    static_assert(sizeof(product_serial_number) >= 20);
+    memcpy(&product_serial_number, identify_ctrl.sn, sizeof(identify_ctrl.sn));
+
+    // Bits 79:72 ASCII representation of “_"
+    product_serial_number[sizeof(identify_ctrl.sn)] = '_';
+
+    // Bits 71:08 ASCII representation of 32 bit Namespace Identifier (NSID)
+    sprintf(&product_serial_number[sizeof(identify_ctrl.sn) + 1], "%08x", nsid);
+
+    // Bits 07:00 ASCII representation of “."
+    product_serial_number[kV1SerialLen - 1] = '.';
+  }
+
+  WriteValue(result, buffer);
+  WriteValue(product_serial_number, buffer.subspan(sizeof(result)));
+}
+}  // namespace
+
+StatusCode InquiryToNvme(absl::Span<const uint8_t> raw_scsi,
+                         nvme::GenericQueueEntryCmd& identify_ns,
+                         nvme::GenericQueueEntryCmd& identify_ctrl,
+                         uint32_t& alloc_len, scsi::LunAddress lun) {
+  scsi::InquiryCommand cmd = {};
+  if (!ReadValue(raw_scsi, cmd)) {
+    DebugLog("Malformed Inquiry Command");
+    return StatusCode::kInvalidInput;
+  };
+
+  alloc_len = cmd.allocation_length;
+
+  // Allocate prp & assign to command
+  uint64_t ns_prp = AllocPages(1);
+  if (!ns_prp) return StatusCode::kFailure;
+
+  identify_ns = nvme::GenericQueueEntryCmd{
+      .opc = static_cast<uint8_t>(nvme::AdminOpcode::kIdentify), .nsid = lun};
```

Comment:
This shouldn't be an issue because NSID is assigned in incrementing order and it's almost inconceivable we'll have to deal with 2^32 namespace IDs. So, casting via truncation should work fine